File: rsj_pick_and_place/src/pick_and_place.py
#!/usr/bin/env python

"""
RSJ pick-and-place demonstration.

For the RSJ 2017 tutorial.

"""

import actionlib
import moveit_commander
import rospy
import sys
import tf2_geometry_msgs
import tf2_ros

from control_msgs.msg import GripperCommandAction, GripperCommandGoal
from geometry_msgs.msg import Point, Pose2D, Pose, Quaternion, Transform, \
                              TransformStamped, Vector3Stamped
from tf.transformations import quaternion_from_euler

GRIPPER_OPEN_WIDTH = 0.1
GRIPPER_CLOSED_WIDTH = 0.0


def setup_arm(scene_frame):
    """Set up the arm for planning."""
    rospy.loginfo('[RSJPickAndPlace] Preparing arm for planning')
    # Set up the arm move group
    arm = moveit_commander.MoveGroupCommander('arm')
    # Allow flexibility in position and orientation to give MoveIt! more
    # freedom in planning
    arm.set_goal_position_tolerance(0.01)
    # arm.set_goal_orientation_tolerance(0.1)
    # arm.set_goal_joint_tolerance(0.05)
    # Allow replanning in case of failing to find a solution
    arm.allow_replanning(True)
    # Set the reference frame
    arm.set_pose_reference_frame(scene_frame)
    # Allow 5 seconds per planning attempt
    #arm.set_planning_time(5)

    # Start the up vertically up
    arm.set_named_target('vertical')
    arm.go()

    return arm


def setup_gripper():
    """Set up the gripper ready for gripping."""
    rospy.loginfo('[RSJPickAndPlace] Preparing gripper for gripping')
    # Create a client for the gripper action
    gripper_client = actionlib.SimpleActionClient(
        '/gripper_command',
        GripperCommandAction)
    # Publish gripper poses
    gripper_pose_pub = rospy.Publisher(
        'gripper_pose',
        Pose,
        queue_size=1)

    # Set the gripper to open
    goal = GripperCommandGoal()
    goal.command.position = GRIPPER_OPEN_WIDTH
    gripper_client.send_goal(goal)
    gripper_client.wait_for_result(rospy.Duration(2))

    return gripper_client, gripper_pose_pub


def make_finger_tip_offset(gripper_rotation, tf_buffer):
    """Calculate the finger tip offset from the gripper frame origin."""
    # Wait for tf data
    while not rospy.is_shutdown():
        rate = rospy.Rate(1)
        try:
            t = tf_buffer.lookup_transform(
                'crane_plus_fixed_finger_tip_link',
                'crane_plus_gripper_link',
                rospy.Time(0))
            break
        except (tf2_ros.LookupException,
                tf2_ros.ConnectivityException,
                tf2_ros.ExtrapolationException):
            rate.sleep()
            continue
    print('Transform is\n' + str(t))
    v = Vector3Stamped(vector=t.transform.translation)
    print('v is\n' + str(v))
    r = TransformStamped(transform=Transform(rotation=gripper_rotation))
    print('r is\n' + str(r))
    trans_vector = tf2_geometry_msgs.do_transform_vector3(v, r).vector
    print('transform is\n' + str(trans_vector))
    return Transform(translation=trans_vector, rotation=Quaternion(w=1))


def open_gripper(gripper_client):
    """Open the gripper."""
    rospy.loginfo('[RSJPickAndPlace] Opening gripper')
    goal = GripperCommandGoal()
    goal.command.position = GRIPPER_OPEN_WIDTH
    gripper_client.send_goal(goal)
    gripper_client.wait_for_result()


def close_gripper(gripper_client):
    """Close the gripper."""
    rospy.loginfo('[RSJPickAndPlace] Closing gripper')
    goal = GripperCommandGoal()
    goal.command.position = GRIPPER_CLOSED_WIDTH
    gripper_client.send_goal(goal)
    gripper_client.wait_for_result()


def do_grasp(
        arm,
        gripper,
        object_pose,
        object_grasp_offset,
        grasp_prepare_z,
        grasp_z,
        gripper_rotation,
        tip_transform):
    """Grasp an object using the arm."""
    # Move to the prepare pose
    grasp_prepare_pose = Pose(Point(
            object_pose.x + tip_transform.translation.x,
            object_pose.y + tip_transform.translation.y - object_grasp_offset,
            grasp_prepare_z + tip_transform.translation.z),
        gripper_rotation)
    rospy.loginfo(
        '[RSJPickAndPlace] Moving arm to grasp-prepare pose:\n' +
        str(grasp_prepare_pose))
    arm.set_pose_target(grasp_prepare_pose)
    if not arm.go():
        rospy.logwarn('[RSJPickAndPlace] Failed to move to pose')
        return False
    # Move to the grasp pose
    grasp_pose = Pose(Point(
            object_pose.x + tip_transform.translation.x,
            object_pose.y + tip_transform.translation.y - object_grasp_offset,
            grasp_z + tip_transform.translation.z),
        gripper_rotation)
    rospy.loginfo(
        '[RSJPickAndPlace] Moving arm to grasp pose:\n' + str(grasp_pose))
    arm.set_pose_target(grasp_pose)
    if not arm.go():
        rospy.logwarn('[RSJPickAndPlace] Failed to move to pose')
        return False
    # Close the gripper
    close_gripper(gripper)
    # Move to the prepare position again
    rospy.loginfo(
        '[RSJPickAndPlace] Moving arm to grasp-prepare pose:\n' +
        str(grasp_prepare_pose))
    arm.set_pose_target(grasp_prepare_pose)
    if not arm.go():
        rospy.logwarn('[RSJPickAndPlace] Failed to move to pose')
        return False
    return True


def do_place(arm, gripper, place_pose):
    """Place an object using the arm."""
    # Move to the place position z+0.05
    place_prepare_pose = place_pose
    place_prepare_pose.position.z += 0.05
    rospy.loginfo(
        '[RSJPickAndPlace] Moving arm to place-prepare pose:\n' +
        str(place_prepare_pose))
    arm.set_pose_target(place_prepare_pose)
    if not arm.go():
        rospy.logwarn('[RSJPickAndPlace] Failed to move to pose')
        return False
    # Move to the place pose
    rospy.loginfo(
        '[RSJPickAndPlace] Moving arm to place pose:\n' + str(place_pose))
    arm.set_pose_target(place_pose)
    if not arm.go():
        rospy.logwarn('[RSJPickAndPlace] Failed to move to pose')
        return False
    # Open the gripper
    open_gripper(gripper)
    return True


def pick_and_place(object_pose, args):
    """Callback for the pose of a new object to pick-and-place."""
    arm, gripper, tf_buffer, place_pose, grasp_prepare_z, grasp_z, \
        gripper_rotation, tip_transform, max_reach, object_grasp_offset = args
    rospy.loginfo('[RSJPickAndPlace] Got new object pose:\n' + str(object_pose))
    if do_grasp(
            arm,
            gripper,
            object_pose,
            object_grasp_offset,
            grasp_prepare_z,
            grasp_z,
            gripper_rotation,
            tip_transform):
        do_place(arm, gripper, place_pose)
    # Make sure the gripper is open even if moving failed
    open_gripper(gripper)
    # Move the arm to an intermediate position
    rospy.loginfo('[RSJPickAndPlace] Moving arm to resting pose')
    arm.set_named_target('resting')
    if not arm.go():
        rospy.logwarn('[RSJPickAndPlace] Failed to move to pose')
    rospy.loginfo(
        '[RSJPickAndPlace] Pick-and-place process finished, ready for '
        'next object')


def main():
    """Main entry point."""
    # Initialise MoveIt! and the node
    moveit_commander.roscpp_initialize(sys.argv)
    rospy.init_node('rsj_pick_and_place')

    # Get the task frames and other configuration from the parameter server
    scene_frame = rospy.get_param('~scene_task_frame', 'world')
    rate = rospy.Rate(rospy.get_param('~execution_rate', 1))
    place_pose = Pose()
    place_pose.position.x = rospy.get_param('~place_pose/x', 0.1)
    place_pose.position.y = rospy.get_param('~place_pose/y', -0.2)
    place_pose.position.z = rospy.get_param('~place_pose/z', 0.05)
    grasp_prepare_z = rospy.get_param('~grasp_prepare_z', 0.07)
    grasp_z = rospy.get_param('~grasp_z', 0.02)
    max_reach = rospy.get_param('~max_reach', 0.23)
    object_grasp_offset = rospy.get_param('~object_grasp_offset', 0.03)
    gripper_rotation = Quaternion(*quaternion_from_euler(
        0,
        rospy.get_param('~gripper_angle', 1.8),
        0))

    # Listen to tf
    tf_buffer = tf2_ros.Buffer()
    tf_listener = tf2_ros.TransformListener(tf_buffer)
    # Make the gripper horizontal in the placing pose
    place_pose.orientation = gripper_rotation
    # Set up the arm
    arm = setup_arm(scene_frame)
    # Set up the gripper
    gripper, gripper_pose_pub = setup_gripper()
    # Calculate the offset from the fingertip to the gripper_link in the scene
    # frame
    tip_transform = make_finger_tip_offset(gripper_rotation, tf_buffer)

    # Subscribe to a topic providing poses of things to pick-and-place
    rospy.Subscriber(
        'objects',
        Pose2D,
        pick_and_place,
        (arm,
         gripper,
         tf_buffer,
         place_pose,
         grasp_prepare_z,
         grasp_z,
         gripper_rotation,
         tip_transform,
         max_reach,
         object_grasp_offset))
    rospy.loginfo('[RSJPickAndPlace] Ready to receive object poses')
    while not rospy.is_shutdown():
        rate.sleep()

    # Shut down MoveIt! cleanly
    moveit_commander.roscpp_shutdown()

    # All done
    return 0


if __name__ == '__main__':
    sys.exit(main())

File: rsj_pick_and_place/src/pick_and_place.cpp
// Copyright 2017 Geoffrey Biggs (geoffrey.biggs@aist.go.jp)

#include <ros/ros.h>

#include <geometry_msgs/Vector3.h>
#include <geometry_msgs/Pose.h>
#include <geometry_msgs/PoseStamped.h>
#include <shape_msgs/SolidPrimitive.h>

#include <actionlib/client/simple_action_client.h>
#include <control_msgs/GripperCommandAction.h>

#include <moveit/move_group_interface/move_group_interface.h>
#include <moveit/planning_scene_interface/planning_scene_interface.h>

#include <moveit_msgs/AttachedCollisionObject.h>
#include <moveit_msgs/CollisionObject.h>

#include <moveit_visual_tools/moveit_visual_tools.h>

#include <string>
#include <vector>


class RSJPickNPlacer {
 public:
  typedef actionlib::SimpleActionClient<control_msgs::GripperCommandAction> GripperClient;

  RSJPickNPlacer()
      : arm_("arm"),
        gripper_("/gripper_command", "true"),
        gripper_closed_width_(0.01),
        gripper_open_width_(0.1),
        prepare_z_(0.1),
        grasp_z_(0.04) {
    scene_frame_ = "base_link";
    place_position_.x = 0.1;
    place_position_.y = -0.2;
    place_position_.z = 0.05;
    SetupPlanningScene();
    SetupArm();
    SetupGripper();
  }

  void DoPickAndPlace(geometry_msgs::Vector3 const& location) {
    ROS_INFO("Received an object to pick-and-place");
    AddBoxToScene(location);
    if (GraspObject(location)) {
      PlaceObject();
    }
    arm_.setNamedTarget("resting");
    arm_.move();
    RemoveBoxFromScene();
    ROS_INFO("Pick-and-place process finished, ready for next object");
  }

 private:
  std::string scene_frame_;
  moveit::planning_interface::PlanningSceneInterface scene_;
  moveit::planning_interface::MoveGroupInterface arm_;

  GripperClient gripper_;
  float gripper_closed_width_;
  float gripper_open_width_;

  geometry_msgs::Vector3 place_position_;
  float prepare_z_;
  float grasp_z_;

  bool DoGripperAction(float width) {
    control_msgs::GripperCommandGoal goal;
    goal.command.position = width;
    gripper_.sendGoal(goal);
    bool finishedBeforeTimeout = gripper_.waitForResult(ros::Duration(30));
    if (!finishedBeforeTimeout) {
      ROS_INFO("Gripper action did not complete");
      return false;
    }
    return true;
  }

  geometry_msgs::Quaternion GripperOrientation() {
    geometry_msgs::Quaternion result;
    result.x = 0;
    result.y = 0.707106781187;
    result.z = 0;
    result.w = 0.707106781187;
    return result;
  }

  void SetupPlanningScene() {
    ROS_INFO("Setting up planning scene");
    // Clear the planning scene
    std::vector<std::string> objs;
    for (auto o: scene_.getObjects()) {
      objs.push_back(o.first);
    }
    for (auto o: scene_.getAttachedObjects()) {
      objs.push_back(o.first);
    }
    scene_.removeCollisionObjects(objs);

    moveit_msgs::CollisionObject table;
    table.header.frame_id = scene_frame_;
    table.id = "table";
    shape_msgs::SolidPrimitive primitive;
    primitive.type = primitive.BOX;
    primitive.dimensions.resize(3);
    primitive.dimensions[0] = 1;
    primitive.dimensions[1] = 1;
    primitive.dimensions[2] = 0.1;
    geometry_msgs::Pose pose;
    pose.position.x = 0;
    pose.position.y = 0;
    pose.position.z = -0.05;
    pose.orientation.w = 1;
    table.primitives.push_back(primitive);
    table.primitive_poses.push_back(pose);
    table.operation = table.ADD;
    std_msgs::ColorRGBA colour;
    colour.b = 0.5;
    colour.a = 1;
    scene_.applyCollisionObject(table, colour);
  }

  void SetupArm() {
    ROS_INFO("Setting up arm for planning");
    arm_.setPoseReferenceFrame(scene_frame_);

    arm_.setNamedTarget("vertical");
    arm_.move();
  }

  void SetupGripper() {
    ROS_INFO("Setting up gripper");
    gripper_.waitForServer();
  }

  void AddBoxToScene(geometry_msgs::Vector3 const& location) {
    moveit_msgs::CollisionObject sponge;
    sponge.header.frame_id = scene_frame_;
    sponge.id = "sponge";
    shape_msgs::SolidPrimitive primitive;
    primitive.type = primitive.BOX;
    primitive.dimensions.resize(3);
    primitive.dimensions[0] = 0.04;
    primitive.dimensions[1] = 0.04;
    primitive.dimensions[2] = 0.031;
    geometry_msgs::Pose pose;
    pose.position.x = 0.17;
    pose.position.y = 0.0;
    pose.position.z = 0.015;
    pose.orientation.w = 1;
    sponge.primitives.push_back(primitive);
    sponge.primitive_poses.push_back(pose);
    sponge.operation = sponge.ADD;
    scene_.applyCollisionObject(sponge);
    ros::Duration(1).sleep();
  }

  void RemoveBoxFromScene() {
    std::vector<std::string> objs;
    objs.push_back("sponge");
    scene_.removeCollisionObjects(objs);
  }

  bool OpenGripper() {
    ROS_INFO("Opening gripper");
    return DoGripperAction(gripper_open_width_);
  }

  bool CloseGripper() {
    ROS_INFO("Closing gripper");
    return DoGripperAction(gripper_closed_width_);
  }

  bool GraspObject(geometry_msgs::Vector3 const& object_position) {
    // Open the gripper
    if (!OpenGripper()) {
      ROS_INFO("Failed to open gripper");
      return false;
    }
    // Move to the approach pose
    geometry_msgs::PoseStamped pose;
    pose.header.frame_id = scene_frame_;
    pose.pose.position.x = object_position.x - 0.02;
    pose.pose.position.y = object_position.y;
    pose.pose.position.z = prepare_z_;
    pose.pose.orientation = GripperOrientation();
    ROS_INFO("Moving arm to grasp-approach pose");
    arm_.setPoseTarget(pose);
    if (!arm_.move()) {
      ROS_INFO("Failed to move to grasp-approach pose");
      return false;
    }
    // Move to the grasp pose
    pose.pose.position.z = grasp_z_;
    ROS_INFO("Moving arm to grasp pose");
    arm_.setPoseTarget(pose);
    if (!arm_.move()) {
      ROS_INFO("Failed to move to grasp pose");
      return false;
    }
    // Close the gripper
    if (!CloseGripper()) {
      ROS_INFO("Failed to close gripper");
      return false;
    }
    // Move to the retreat pose
    pose.pose.position.z = prepare_z_;
    ROS_INFO("Moving arm to grasp-retreat pose");
    arm_.setPoseTarget(pose);
    if (!arm_.move()) {
      ROS_INFO("Failed to move to grasp-retreat pose");
      return false;
    }
    return true;
  }

  bool PlaceObject() {
    // Move to the approach pose
    geometry_msgs::PoseStamped pose;
    pose.header.frame_id = scene_frame_;
    pose.pose.position.x = place_position_.x;
    pose.pose.position.y = place_position_.y;
    pose.pose.position.z = prepare_z_;
    pose.pose.orientation = GripperOrientation();
    ROS_INFO("Moving arm to place-approach pose");
    arm_.setPoseTarget(pose);
    if (!arm_.move()) {
      ROS_INFO("Failed to move to place-approach pose");
      return false;
    }
    // Move to the place pose
    pose.pose.position.z = place_position_.z;
    ROS_INFO("Moving arm to place pose");
    arm_.setPoseTarget(pose);
    if (!arm_.move()) {
      ROS_INFO("Failed to move to place pose");
      return false;
    }
    // Open the gripper
    if (!OpenGripper()) {
      ROS_INFO("Failed to open gripper");
      return false;
    }
    // Move to the retreat pose
    pose.pose.position.z = prepare_z_;
    ROS_INFO("Moving arm to place-retreat pose");
    arm_.setPoseTarget(pose);
    if (!arm_.move()) {
      ROS_INFO("Failed to move to place-retreat pose");
      return false;
    }
    return true;
  }
};

int main(int argc, char **argv) {
  ros::init(argc, argv, "rsj_pick_and_place");
  ros::NodeHandle nh;
  ros::AsyncSpinner spinner(1);
  spinner.start();

  RSJPickNPlacer pnp;
  geometry_msgs::Vector3 obj;
  obj.x = 0.2;
  obj.y = 0;
  obj.z = 0;
  pnp.DoPickAndPlace(obj);

  ros::shutdown();
  return 0;
}
